src/bin/copy.c

```diff
@@ -79,18 +83,87 @@ static char *getuser(void)
 	return pw->pw_name;
 }
 
+/*
+ * If UNIX user is in UNIX group of directory containing file,
+ * return gid, otherwise -1
+ *
+ * E.g., writing to /cfg/foo, where /cfg is owned by root:wheel,
+ * should result in the file being owned by $LOGNAME:wheel with
+ * 0660 perms for other users in same group.
+ */
+static gid_t in_group(const char *user, const char *fn, gid_t *gid)
+{
+	char dir[strlen(fn) + 2];
+	const struct passwd *pw;
+	int i, num = 0, rc = 0;
+	struct stat st;
+	gid_t *groups;
+	char *ptr;
+
+	pw = getpwnam(user);
+	if (!pw)
+		return 0;
+
+	strlcpy(dir, fn, sizeof(dir));
+	ptr = strrchr(dir, '/');
+	if (ptr)
+		*ptr = 0;
+	else
+		strlcpy(dir, ".", sizeof(dir));
+
+	if (stat(dir, &st))
+		return 0;
+
+	getgrouplist(user, pw->pw_gid, NULL, &num);
+
+	groups = malloc(num * sizeof(gid_t));
+	if (!groups) {
+		perror("in_group() malloc");
+		return 0;
+	}
+
+	getgrouplist(user, pw->pw_gid, groups, &num);
```

Comment:
Isn't there's a theoretical race here, if the user is added to a group after the last call to getgrouplist()? If so, perhaps it's worth wrapping this in an if (..... != -1).

Reply:
There is, wrapping it with a `realloc()` is probably a good idea to be safer.